src/xGadgetron/cGadgetron/include/sirf/Gadgetron/TrajectoryPreparation.h

```diff
@@ -49,26 +49,108 @@ namespace sirf{
 * The interface provides set_trajectory() which populates this data field depending on the implementation.
 */
+template <uint16_t D>
 class TrajectoryPreparation{
 
 public:
+    typedef typename std::array<float, D> TrajPointType;
+    typedef typename std::vector<TrajPointType> TrajPointSet;
+
     TrajectoryPreparation(){}
-    virtual void set_trajectory(sirf::MRAcquisitionData& mr_acq) =0;
     
-protected:
+    virtual void set_trajectory(MRAcquisitionData& mr_acq)
+    {
+        update_acquisitions_info(mr_acq);
+
+        for(size_t ia=0; ia<mr_acq.number(); ++ia)
+        {
+            ISMRMRD::Acquisition acq;
+            mr_acq.get_acquisition(ia, acq);
+            this->set_acquisition_trajectory(acq);
+            mr_acq.set_acquisition(ia, acq);
+        }
+    }        
+
+    virtual TrajPointSet get_trajectory(const MRAcquisitionData& mr_acq, const bool unique_kz=true)
+    {
+        if(mr_acq.number() <= 0){
+            throw std::runtime_error("Please pass a non-empty container.");
+        }
+        
+        ISMRMRD::Acquisition acq;
+        mr_acq.get_acquisition(0, acq);
+
+        if( acq.trajectory_dimensions() != D){
+            throw std::runtime_error("Please give Acquisition with a the correct dimensionality if you want to use it here.");
+        }
+        
+        std::vector<size_t> kspace_dims;
+        mr_acq.get_kspace_dimensions(kspace_dims);
+
+        size_t const Nz = mr_acq.acquisitions_info().get_IsmrmrdHeader().encoding[0].encodedSpace.matrixSize.z;
+        size_t const num_traj_points = unique_kz ? mr_acq.number()*kspace_dims[0]/Nz 
+                                                 : mr_acq.number()*kspace_dims[0];
+
+        TrajPointSet traj;
+        
+        if(unique_kz)
+            traj.resize(num_traj_points);
+            
+        for(int ia=0; ia<mr_acq.number(); ++ia)
+        {
+            mr_acq.get_acquisition(ia, acq);
+            
+            if(unique_kz)
+            {
+                if(acq.idx().kspace_encode_step_2 != 0)
+                    continue;
+                    
+                get_ky_sorted_trajectory(traj, acq);
+            }
+            else
+                append_to_trajectory(traj, acq);
+        }
+
+        return traj;
+    }
 
-    void update_acquisitions_info(sirf::MRAcquisitionData& mr_acq);
 
+protected:
     ISMRMRD::Encoding kspace_encoding_;
     ISMRMRD::TrajectoryType traj_type_;
 
-    uint16_t traj_dim_;
-    virtual void set_acquisition_trajectory(ISMRMRD::Acquisition& acq) const =0;
-    virtual std::vector<float> calculate_trajectory(ISMRMRD::Acquisition& acq) const =0;
+    virtual void set_acquisition_trajectory(ISMRMRD::Acquisition& acq) const
+    {
+        acq.resize(acq.number_of_samples(),acq.active_channels(), D);
+        TrajPointSet acq_traj = this->calculate_trajectory(acq);
+        acq.setTraj(&(acq_traj[0][0]));
+    }
+    virtual void update_acquisitions_info(sirf::MRAcquisitionData& mr_acq)
```

Comment:
doxygen here as well (what's the difference with the `set_acquisition_trajectory` (aside from the input type)? Should they be called similarly? (why `set` and `update`?)

Reply:
Renamed it